src/main/java/de/tum/cit/aet/artemis/programming/service/localci/SharedQueueManagementService.java

```diff
@@ -323,4 +334,123 @@ public Page<BuildJob> getFilteredFinishedBuildJobs(FinishedBuildJobPageableSearc
+        }
+
+        if (jobsQueuedBefore.size() < agentsAvailabilities.size()) {
+            return now.plusSeconds(agentsAvailabilities.get(jobsQueuedBefore.size()));
+        }
+        else {
+            return now.plusSeconds(calculateNextJobQueueDuration(agentsAvailabilities, jobsQueuedBefore));
+        }
+    }
+
+    private String getIdOfQueuedJobFromParticipation(long participationId) {
+        var participationBuildJobIds = getQueuedJobs().stream().filter(job -> job.participationId() == participationId).map(BuildJobQueueItem::id).toList();
+        if (participationBuildJobIds.isEmpty()) {
+            return null;
+        }
+        return participationBuildJobIds.getLast();
+    }
+
+    private Long calculateNextJobQueueDuration(List<Long> agentsAvailabilities, List<BuildJobQueueItem> jobsQueuedBefore) {
+        PriorityQueue<Long> agentAvailabilitiesQueue = new PriorityQueue<>(agentsAvailabilities);
+        for (BuildJobQueueItem job : jobsQueuedBefore) {
+            Long agentRemainingTimeObj = agentAvailabilitiesQueue.poll();
+            long agentRemainingTime = agentRemainingTimeObj == null ? 0 : agentRemainingTimeObj;
+            agentRemainingTime = Math.max(0, agentRemainingTime);
+            agentAvailabilitiesQueue.add(agentRemainingTime + job.jobTimingInfo().estimatedDuration());
+        }
+        Long agentRemainingTimeObj = agentAvailabilitiesQueue.poll();
+        return agentRemainingTimeObj == null ? 0 : agentRemainingTimeObj;
+    }
+
+    private long buildJobRemainingDuration(BuildJobQueueItem buildJob, ZonedDateTime now) {
+        ZonedDateTime estimatedCompletionDate = buildJob.jobTimingInfo().estimatedCompletionDate();
+        if (estimatedCompletionDate == null) {
+            return 0;
+        }
+        if (estimatedCompletionDate.isBefore(now)) {
+            return 0;
+        }
+        return Duration.between(now, estimatedCompletionDate).toSeconds();
+
+    }
+
+    private class BuildAgentListener
+            implements EntryAddedListener<String, BuildAgentInformation>, EntryRemovedListener<String, BuildAgentInformation>, EntryUpdatedListener<String, BuildAgentInformation> {
+
+        @Override
+        public void entryAdded(com.hazelcast.core.EntryEvent<String, BuildAgentInformation> event) {
+            log.debug("Build agent added: {}", event.getValue());
+            getBuildAgentsCapacity();
+        }
+
+        @Override
+        public void entryRemoved(com.hazelcast.core.EntryEvent<String, BuildAgentInformation> event) {
+            log.debug("Build agent removed: {}", event.getOldValue());
+            getBuildAgentsCapacity();
+        }
+
+        @Override
+        public void entryUpdated(com.hazelcast.core.EntryEvent<String, BuildAgentInformation> event) {
+            log.debug("Build agent updated: {}", event.getValue());
+            getBuildAgentsCapacity();
+        }
+    }
+
+    private void getBuildAgentsCapacity() {
```

Comment:
As this method does not return (`get`) anything but works purely using side-effects (setting class variables), I'd call this method rather `updateBuildAgentCapacity` or `updateCapacityForBuildAgents`